Speedup BetaGeoModel tests

ricardoV94 opened the pull request:
This PR removes the costly fitting from the setup class. Even running a small test like `test_model_repr` would take around one minute as it would have to wait for the useless fitting (in the context of that test) to be done.

Also simplified the patching of the lifetimes reference model. It's enough to override `params_` after initializing it.

Finally added a `pytest.mark.slow` for the `test_model_convergence` so it doesn't run by default (it does run in our CI here). This way the tests take just ~2 seconds on my slow laptop.

ColtAllen commented:
To speed this up further, should the model fitting in `setup_class()` be done with MAP? `test_model_convergence()` is already covering a MCMC fit.

ColtAllen commented on tests/clv/models/test_beta_geo.py:
```diff
@@ -147,7 +105,7 @@ def test_model_convergence(self, N, rtol):
             T=T,
         )
-        model.fit(chains=1, progressbar=False, random_seed=self.rng)
+        model.fit(chains=1, progressbar=False, random_seed=rng)
```
Should a MAP fit also be added here for the CI?

ColtAllen commented on tests/clv/models/test_beta_geo.py:
```diff
@@ -160,63 +118,133 @@ def test_expected_probability_alive(self):
             rtol=0.05,
         )
 
     def test_expected_num_purchases(self):
```
What's the justification for combining these two tests? These are two separate methods with different input and output dimensions.

ricardoV94 replied:
They both require building the same type of values, patching the pymc-makerting and lifetimes model, and they are both tested against the same lifetimes method. There was a TODO before suggesting they be merged already, probably for similar reasoning

ricardoV94 replied:
Done

ColtAllen reviewed (commented):
`BetaGeoModel` is being redefined in every test. Would it make more sense to define it only once in `setup_class()` as was the case previously? Same with `lifetimes.BetaGeoFitter` as it could be compared against in `test_expected_probability_alive`.  This would also simplify required changes if we decide to implement https://github.com/pymc-labs/pymc-marketing/issues/100.

larryshamalama commented on tests/clv/models/test_beta_geo.py:
```diff
@@ -160,63 +118,133 @@ def test_expected_probability_alive(self):
+        rng = np.random.default_rng(152)
+
+        N = 100
+        # Almost deterministic p = .02
```
What does an almost deterministic dropout probability here check?

ricardoV94 replied:
That the probability of dropout is stable at a value that should induce a nice `p_alive` ~ 0.5, nothing too low nor too high

ricardoV94 replied:
Added a comment

larryshamalama commented on tests/clv/models/test_beta_geo.py:
```diff
@@ -160,63 +118,133 @@ def test_expected_probability_alive(self):
+            a_prior=pm.Normal.dist(a, 1e-3),
+            b_prior=pm.Normal.dist(b, 1e-3),
+            alpha_prior=pm.Normal.dist(alpha, 1e-3),
+            r_prior=pm.Normal.dist(r, 1e-3),
```
Why the informative priors here?

ricardoV94 replied:
For a fake posterior based on prior_predictive sampling around the true values

ricardoV94 replied:
Replaced by direct sampling with numpy when creating the fake fit

larryshamalama reviewed (commented):
Looks great, thanks @ricardoV94! The initial solution of having the model fit without the `pytest.mark.slow` was a little naive and due for a speed up. I just added some questions regarding some parameter choices in the tests.

ricardoV94 commented:
> `BetaGeoModel` is being redefined in every test. Would it make more sense to define it only once in `setup_class()` as was the case previously? Same with `lifetimes.BetaGeoFitter` as it could be compared against in `test_expected_probability_alive`. This would also simplify required changes if we decide to implement #100.

No, I think we should create it every time, because we usually monkey-patch the fit result and we don't want to be operating on the same stateful object (i.e., erasing a previous fit).

It was a conscious choice not to create a class-level model in the other model tests, like the gamma-gamma.

ricardoV94 commented:
I set this as draft as it includes the commit from #198 

Otherwise I think I addressed your review comments @larryshamalama and @ColtAllen. Let me know if there's anything else.

larryshamalama reviewed (approved):
Looks good to me :)
